### PR TITLE
Improve numerical stability of GroupNorm

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary: Improve numerical stability of GroupNorm

Test Plan: buck test mode/dev-nosan //caffe2/test:nn -- "GroupNorm"

Differential Revision: D27414438

There are two main changes in this PR:
1. Use Welford algorithm to compute main and variance.
2. Use cascade summation when computing sum over input for both mean and variance.

Some benchmark results.

input_size = [256, 1024, 28, 28], group = 8

for single thread CPU run average time change is 510.91ms -> 602.70ms.
for CUDA run average time change is 12.20ms -> 12.47ms.

### CODE REVIEW

File: aten/src/ATen/native/cpu/moments_utils.h [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+// numerical stability.
+// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+// https://en.wikipedia.org/wiki/Pairwise_summation
+template <typename T>
+std::pair<T, T> RowwiseMoments(const T* X, int64_t N) {
+  using Vec = vec256::Vec256<T>;
+
+  constexpr int64_t kVecSize = Vec::size();
+  constexpr int64_t kChunkSize = 32;
+  constexpr int64_t kMaxDepth = 64;
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/moments_utils.h:63]: max depth of 64 seems like an overkill (given that it is then used as the size for the m0_stk, m1_stk, m2_stk variables)? In the reduction PR #39516 there were just 4 levels and that seemed enough for pretty large tensors.

File: aten/src/ATen/native/cpu/SumKernel.cpp [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
-#include <ATen/native/cpu/Reduce.h>
-#include <c10/util/llvmMathExtras.h>
 
 #include <algorithm>
 
+#include <ATen/Dispatch.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cpu/Reduce.h>
+#include <ATen/native/cpu/utils.h>
+#include <c10/util/llvmMathExtras.h>
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/SumKernel.cpp:9]: nit: looks like llvmMathExtras are not needed here, only by cpu/utils.h?

File: aten/src/ATen/native/cuda/block_reduce.cuh [hunk 3]
```diff
@@ -34,6 +36,55 @@ __inline__ __device__ T BlockReduceSum(T val, T* shared) {
   return val;
 }
 
+template <typename T>
+__inline__ __device__ thrust::tuple<int64_t, T, T> WarpReduceMoments(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/block_reduce.cuh:40]: is there any reason you are not using WelfordOps.combine from SharedReduceOps.h here?

File: torch/testing/_internal/common_nn.py [hunk 4]
```diff
@@ -1674,6 +1674,19 @@ def fractional_max_pool3d_test(test_case):
...
+        constructor_args=(3, 12, 1e-3),
+        cpp_constructor_args='torch::nn::GroupNormOptions(3, 6).eps(1e-3)',
+        input_size=(4, 12, 28, 28),
+        cudnn=True,
+        check_eval=True,
+        check_bfloat16=True,
+        desc='2d_no_affine_large_feature_forward',
+        check_forward_only=True,
+        reference_fn=lambda x, *_: torch.nn.functional.layer_norm(
+            x.view(4, 3, -1), (4 * 28 * 28,), eps=1e-3).view(4, 12, 28, 28)
```
Comment (ngimel) [inline torch/testing/_internal/common_nn.py:1688]: is this size enough to observe better accuracy (i.e. will it fail without this PR?). Also, IIRC you never landed LayerNorm better accuracy PRs, so using LayerNorm as reference could be inaccurate?

File: aten/src/ATen/native/cpu/moments_utils.h (hunk 1 above)
Comment (xiaomengy) [reply aten/src/ATen/native/cpu/moments_utils.h]: Personally I think it will be more reliable to keep a fixed chunk size, especially for the case in var because there is square op for the input. I changed the chunk size to 128 which is the same as numpy below. And also make the kMaxDepth as a parameter for the template function. Then if the size is small, the SmallVector will be smaller to improve the register efficiency.
https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/core/src/umath/loops_utils.h.src#L60

File: aten/src/ATen/native/cpu/SumKernel.cpp (hunk 2 above)
Comment (xiaomengy) [reply aten/src/ATen/native/cpu/SumKernel.cpp]: Done.

File: aten/src/ATen/native/cuda/block_reduce.cuh (hunk 3 above)
Comment (xiaomengy) [reply aten/src/ATen/native/cuda/block_reduce.cuh]: Changed to WelfordOps

File: torch/testing/_internal/common_nn.py (hunk 4 above)
Comment (xiaomengy) [reply torch/testing/_internal/common_nn.py]: The original purpose of this test is to make sure we cover the test cases larger than the chunksize. It is a little hard to have a reliable test case for the numerical issue. If the input size is large, the test will be pretty slow. I removed the layer_norm and used torch.mean and torch.std for reference_fn.

**Review by ngimel — COMMENTED** [review]
This looks good!
Can you please post benchmarks, both for cuda and cpu?
Also, consider implementing multi-stage cascade summation, like in original `sum` PR or #55841 to limit number of levels (64 is a bit too much), but it's not a blocker, PR can go in as is.

File: aten/src/ATen/native/cpu/moments_utils.h (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/native/cpu/moments_utils.h]: I chose a fixed number of levels because I expect the compiler to do a lot better at allocating registers for a fixed number of accumulators. I don't think a dynamically-sized array of vector is even possible to keep in registers.

For four levels with a chunk size of 16, and a vector size of 8 (for float), you can sum up to 2^19 or .5 million elements. With 5 levels and a chunk size of 16, you can sum 2^63 elements i.e. more floats than you can possibly address with 64-bits.

With 64 levels and a chunk size of 16, you could count every atom in the universe.

File: aten/src/ATen/native/cpu/moments_utils.h [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    for (int64_t j = 0; j < m0; ++j) {
+      const Vec x_vec = Vec::loadu(X_ptr + j * kVecSize);
+      const Vec delta_vec = x_vec - m1_vec;
+      const Vec c_vec = Vec(T(1) / static_cast<T>(j + 1));
+      m1_vec += delta_vec * c_vec;
+      m2_vec += delta_vec * (x_vec - m1_vec);
+    }
+    AddMomentsVec(m0, m1_vec, m2_vec, m0_stk[0], m1_stk[0], m2_stk[0]);
+    int64_t mask = i + 1;
+    for (int64_t j = 1; j < depth && (mask & 1) == 0; ++j) {
```
Comment (peterbell10) [inline aten/src/ATen/native/cpu/moments_utils.h:84]: This isn't right. Each level of the accumulator should be the sum of `kChunkSize` elements on the level below.

File: aten/src/ATen/native/SharedReduceOps.h [hunk 6]
```diff
@@ -80,8 +80,20 @@ struct WelfordData {
   scalar_t m2;
   index_t n;
   combine_t nf;
-  C10_HOST_DEVICE WelfordData() : mean(0), m2(0), n(0), nf(0)  {}
-  C10_DEVICE WelfordData(scalar_t mean, scalar_t m2, index_t n, combine_t nf) : mean(mean), m2(m2), n(n), nf(nf) {}
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  // Shared memory requires no initialization.
+  C10_HOST_DEVICE WelfordData() = default;
```
[the remark this answers is not in the document]
Comment (xiaomengy) [reply aten/src/ATen/native/SharedReduceOps.h:86]: Yes, but if we add dynamic initialization in ctor here, there will be a warning "dynamic initialization is not supported for a function-scope static __shared__ variable within a __device__/__global__ function".

File: aten/src/ATen/native/cpu/moments_utils.h (hunk 5 above)
Comment (xiaomengy) [reply aten/src/ATen/native/cpu/moments_utils.h:84]: This is the correct formula. Here I only do the chunk_size reduction on the leaf node. The reduction tree is a binary tree. The pseudo code is the same as numpy's implementation below. Here I think it will be more reliable to do the exact pairwise summation for var since it is actually the 2nd order moment, which will enlarge the numeric issue for summation.
https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/core/src/umath/loops_utils.h.src#L135

**Review by xiaomengy — COMMENTED** [review]
Here I think it will be more reliable to do the exact pairwise summation for var since it is actually the 2nd order moment, which will enlarge the numeric issue for summation. So I didn't keep a fixed size of depth. Since there is only the chunk-size reduction on the leaf node, so the actual max depth should be 63 - (4 + 3) = 56. So I think the maximum of 64 should be reasonable. Though in practice we will never use that size.

File: aten/src/ATen/native/SharedReduceOps.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/SharedReduceOps.h:86]: Sorry for delay, I'm still uneasy about this initialization, otherwise PR looks good.

[the remark this answers is not in the document]
Comment (xiaomengy) [reply aten/src/ATen/native/SharedReduceOps.h:86]: Actually there are some discussions for this before. Not sure if there is any better way to handle this.
https://github.com/pytorch/pytorch/pull/13967

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/native/SharedReduceOps.h:86]: Quoting from our compiler team for a similar issue:

```
__shared__ variables cannot be initialized.
```
...
and for the case of dynamic initialization:
```
the compiler reports a warning (rather than an error, for legacy reasons). The program will build but will not execute correctly. The warning reported is:

j5.cu(6): warning: dynamic initialization is not supported for a function-scope static __shared__ variable within a __device__/__global__ function
```

Can we keep the default ctor unchanged as
```
C10_HOST_DEVICE WelfordData() : mean(0), m2(0), n(0), nf(0)  {}
```

and inside the kernel, do
```CUDA
__shared__ char val_shared_[C10_WARP_SIZE * sizeof(WelfordType)];
WelfordType *val_shared = reinterpret_cast<WelfordType *>(val_shared_);
if (threadIdx.x < 32) {
  val_shared[threadIdx.x] = WelfordType();
}
__syncthreads();
```

[the remark this answers is not in the document]
Comment (xiaomengy) [reply aten/src/ATen/native/SharedReduceOps.h]: Thanks for the advice.

File: aten/src/ATen/native/cuda/group_norm_kernel.cu [hunk 7]
```diff
@@ -32,29 +32,36 @@ __global__ void RowwiseMomentsCUDAKernel(
...
-    sum2 = cuda_utils::WarpReduceSum<T_ACC>(sum2);
+    val = cuda_utils::WarpReduce(val, welford_op);
   } else {
-    __shared__ T_ACC m_shared[C10_WARP_SIZE];
-    __shared__ T_ACC v_shared[C10_WARP_SIZE];
-    sum1 = cuda_utils::BlockReduceSum<T_ACC>(sum1, m_shared);
-    sum2 = cuda_utils::BlockReduceSum<T_ACC>(sum2, v_shared);
+    // There will be a warning if we declare a __shared__ WelfordType array.
+    // https://github.com/pytorch/pytorch/pull/13967
+    __shared__ char val_shared[C10_WARP_SIZE * sizeof(WelfordType)];
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/group_norm_kernel.cu:51]: this can potentially result in misaligned access, so please use 
```
__shared__ std::aligned_storage<sizeof(WelfordType), alignof(WelfordType)>::type val_shared
```

File: torch/testing/_internal/common_nn.py [hunk 8]
```diff
@@ -1647,6 +1647,20 @@ def fractional_max_pool3d_test(test_case):
         desc='2d_affine_large_feature',
         test_cpu=False,
     ),
+    dict(
+        module_name='GroupNorm',
+        constructor_args=(3, 48, 0.0),
+        cpp_constructor_args='torch::nn::GroupNormOptions(3, 48).eps(0.0)',
+        input_fn=lambda: torch.rand(2, 48, 15, 15),
```
Comment (ngimel) [inline torch/testing/_internal/common_nn.py:1654]: Do you think this size is enough to check accuracy?

**Review by ngimel — COMMENTED** [review]
This looks good, I have a minor comment about alignment.

File: torch/testing/_internal/common_nn.py (hunk 8 above)
Comment (xiaomengy) [reply torch/testing/_internal/common_nn.py]: A smoke test case for large input added.

File: aten/src/ATen/native/cuda/group_norm_kernel.cu (hunk 7 above)
Comment (xiaomengy) [reply aten/src/ATen/native/cuda/group_norm_kernel.cu]: Thanks for the suggestions!

File: test/test_nn.py [hunk 9]
```diff
@@ -5760,6 +5760,38 @@ def test_RNN_cell_forward_hidden_size(self):
                      nn.GRUCell(*cell_shared_param)):
             self.assertRaises(Exception, lambda: cell(input, hx))
 
+    def test_GroupNorm_numeric(self):
```
Comment (ngimel) [inline test/test_nn.py:5763]: can you move this test into device-generic part of the test suite (TestNNDeviceType at line ~11568) so that you'd have device argument to handle cuda testing?
How long does the cpu test take? Should it be decorated with @slowTest?

Comment (xiaomengy) [reply test/test_nn.py]: Done.

**Review by ngimel — APPROVED** [review]
